### PR TITLE
feat: sign and broadcast Stacks transactions

### PR DESCRIPTION
**Pull request by djordon**
## Description

Closes https://github.com/stacks-network/sbtc/issues/616.

Note that we purposefully skip validation for now, but it should be relatively straightforward to hook up.

## Changes

* Make sure the client types are `&'static`, which they all are. This was needed for borrow checker.
* Send the transaction ID with the `StacksTransactionSignRequest`. We can probably remove the digest from that type, we will do that later.
* Make sure we have a loop for signing stacks transactions.
* Submit the signed transaction to the stacks node

This PR needs some polish and some tests.

## Testing Information

The main test that is needed is that the transaction coordinator broadcasts a signed `complete-deposit` transaction for a deposit request with a sweep transaction on bitcoin. 

## Checklist:

- [ ] I have performed a self-review of my code

### CODE REVIEW

**Comment by cylewitruk** [comment]
> Note that we purposefully skip validation for now, but it should be relatively straightforward to hook up.

Is there a separate ticket for this?

**Comment by djordon** [comment]
> > Note that we purposefully skip validation for now, but it should be relatively straightforward to hook up.
> 
> Is there a separate ticket for this?

No not yet. There are a few TODOs in this PR that I need to turn into outstanding tickets, this was just one of them.

File: signer/src/message.rs [hunk 1]
```diff
@@ -165,6 +165,8 @@ pub struct StacksTransactionSignRequest {
     /// It's essentially a hash of the contract call struct, the nonce, the
     /// tx_fee and a few other things.
     pub digest: [u8; 32],
+    /// The transaction ID of the associated contract call transaction.
+    pub txid: blockstack_lib::burnchains::Txid,
```
Comment (matteojug) [inline signer/src/message.rs:169]: Is this docstring/type correct? What associated contract call tx (on bitcoin?) are we referring to?

File: signer/src/transaction_coordinator.rs [hunk 2]
```diff
@@ -288,20 +291,61 @@ where
...
+            // If we fail here, then we need to decrement the nonce.
+            let process_request_fut =
+                self.process_sign_request(sign_request, chain_tip, multi_tx, &wallet);
+
+            match process_request_fut.await {
+                Ok(txid) => {
+                    tracing::info!(%txid, "successfully submitted complete-deposit transaction")
+                }
+                Err(error) => {
+                    tracing::warn!(%error, "could process the sign request for deposit request");
```
Comment (matteojug) [inline signer/src/transaction_coordinator.rs:315]:
```suggestion
                    tracing::warn!(%error, "could not process the sign request for deposit request");
```

File: signer/src/transaction_coordinator.rs [hunk 3]
```diff
@@ -288,20 +291,61 @@ where
...
         Ok(())
     }
 
+    /// Sign and broadcast the stacks transaction
+    ///
+    /// If we fail to sign the transaction for some reason, we decrement
+    /// the nonce by one, and try the next transaction. This is not a fatal
+    /// error, since we could fail to sign the transaction because someone
+    /// else is now the coordinator, and all of the signers are now
+    /// ignoring us.
```
Comment (matteojug) [inline signer/src/transaction_coordinator.rs:330]: nit: move this part above? It seems out of place here, since we are not decrementing the nonce (here)

**Review by matteojug — COMMENTED** [review]
Just some nits, so far LGTM.

File: signer/src/message.rs (hunk 1 above)
Comment (djordon) [reply signer/src/message.rs:169]: Yeah I believe this is correct. Maybe we should use a type alias or the internal `StacksTxId` type here to make that clear, but `blockstack_lib::burnchains::Txid` is the type for transaction IDs on stacks. See:
https://github.com/stacks-network/stacks-core/blob/2e67fa900b5a1ef9f7268529b9c2d1fdcc72c5d4/stackslib/src/chainstate/stacks/transaction.rs#L824-L829
and in that same file:
https://github.com/stacks-network/stacks-core/blob/2e67fa900b5a1ef9f7268529b9c2d1fdcc72c5d4/stackslib/src/chainstate/stacks/transaction.rs#L34-L36
.

File: signer/src/transaction_coordinator.rs (hunk 2 above)
Comment (djordon) [reply signer/src/transaction_coordinator.rs]: Done in https://github.com/stacks-network/sbtc/pull/617/commits/7cdea8d577b0e2a7d7ed9335e26d38e87252a0e9

File: signer/src/transaction_coordinator.rs (hunk 3 above)
Comment (djordon) [reply signer/src/transaction_coordinator.rs]: Done in https://github.com/stacks-network/sbtc/pull/617/commits/7cdea8d577b0e2a7d7ed9335e26d38e87252a0e9

File: signer/src/transaction_signer.rs [hunk 4]
```diff
@@ -435,12 +443,14 @@ where
...
-        ctx: &impl Context,
-        request: &message::StacksTransactionSignRequest,
+    async fn assert_valid_stacks_tx_sign_request(
+        &self,
+        request: &StacksTransactionSignRequest,
         chain_tip: &model::BitcoinBlockHash,
     ) -> Result<(), Error> {
+        if true {
+            return Ok(());
+        }
```
Comment (matteojug) [inline signer/src/transaction_signer.rs:453]: Do we have a TODO/issue for this?

**Review by matteojug — APPROVED** [review]

File: signer/src/testing/context.rs [hunk 5]
```diff
@@ -162,10 +162,10 @@ impl<Storage, Bitcoin, Stacks>
 
 impl<Storage, Bitcoin, Stacks, Emily> Context for TestContext<Storage, Bitcoin, Stacks, Emily>
 where
-    Storage: DbRead + DbWrite + Clone + Sync + Send,
-    Bitcoin: BitcoinInteract + Clone + Send + Sync,
-    Stacks: StacksInteract + Clone + Send + Sync,
-    Emily: EmilyInteract + Clone + Send + Sync,
+    Storage: DbRead + DbWrite + Clone + Sync + Send + 'static,
```
Comment (cylewitruk) [inline signer/src/testing/context.rs:165]: What was it that made these all need to be 'static?

**Review by cylewitruk — APPROVED** [review]
Looks good in general

File: signer/src/transaction_signer.rs (hunk 4 above)
Comment (djordon) [reply signer/src/transaction_signer.rs:453]: Oh we do, in the line below.

**Comment by aldur** [comment]
@djordon can you lay out the test plan so that @matteojug can take it?

File: signer/src/testing/context.rs (hunk 5 above)
Comment (djordon) [reply signer/src/testing/context.rs:165]: I was getting some compilation error without the `'static` bound on the Stacks client. I'm not sure if I got the error with all of them or not (certainly not with Emily). I did not revisit whether they could be removed after the code was finished though, so it's possible that we can remove the `'static` bounds now.

**Pull request merged**